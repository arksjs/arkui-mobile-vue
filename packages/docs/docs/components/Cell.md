# Cell 单元格

单元格为列表中的单个展示项。

<CodeDemo name="Cell">

<<< @/../../demo/src/components/Show/Cell/index.vue

</CodeDemo>

## Import

```js
import { TaCell } from 'tantalum-ui-mobile'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

### Import Type

组件导出的类型定义：

```ts
import type { CellArrowDirection } from 'tantalum-ui-mobile'
```

## Props

| 属性            | 类型               | 默认值  | 必填 | 说明                                                                |
| --------------- | ------------------ | ------- | ---- | ------------------------------------------------------------------- |
| label           | string             |         | 是   | 左侧文字                                                            |
| description     | string             |         | 否   | 左侧附加描述文字                                                    |
| content         | string             |         | 否   | 右侧文字                                                            |
| required        | boolean            | false   | 否   | true 在左侧文字边上会展示一个 红色`*`箭头                           |
| clickable       | boolean            | false   | 否   | 是否开启点击反馈                                                    |
| is-link         | boolean            | false   | 否   | 是否展示右侧箭头并开启点击反馈                                      |
| arrow-direction | CellArrowDirection | 'right' | 否   | isLink=true 时展示的箭头方向，可选值：'right', 'up', 'down', 'left' |

## Slots

### 内容区（#default）

```vue
<ta-cell label="右侧图标">
  <ta-icon icon="CloseOutlined"></ta-icon>
</ta-cell>
```

注：添加 slot 后 `content` prop 属性失效。
