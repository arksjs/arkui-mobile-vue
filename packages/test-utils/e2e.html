<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
