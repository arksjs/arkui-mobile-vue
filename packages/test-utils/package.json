{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "UI test-utils",
  "main": "dist/index.js",
  "private": true
}
