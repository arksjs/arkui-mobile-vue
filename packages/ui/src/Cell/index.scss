@import '../style/var.scss';

.ta-cell {
  --ta-label-color: #{$title-color};
  --ta-content-color: #{$font-color};

  display: block;
  min-height: 48px;
  background-color: $white-color;
  position: relative;
  width: 100%;

  &:not(:last-child)::after {
    content: '';
    margin-left: 16px;
    width: calc(100% - 16px);
  }

  &.has--icon::after {
    margin-left: 48px;
    width: calc(100% - 48px);
  }

  &_header {
    padding: 8px 16px;
    box-sizing: border-box;
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-wrap: wrap;
  }

  &_content {
    flex-grow: 1;
    flex-shrink: 0;
    font-size: 17px;
    color: var(--ta-content-color);
    min-height: 32px;
    line-height: 32px;
    display: flex;
    align-items: center;
    max-width: 100%;
  }

  &_label {
    flex-shrink: 0;
    height: 32px;
    display: -webkit-box;
    -webkit-line-clamp: 1;
    -webkit-box-orient: vertical;
    overflow: hidden;
    line-height: 32px;
    color: var(--ta-label-color);
    font-size: 17px;
    padding-right: 12px;

    + .ta-cell_content {
      justify-content: flex-end;
    }
  }

  &_icon {
    margin: 0 12px 0 0;

    > .ta-icon {
      --ta-icon-size: 20px;
      --ta-icon-color: var(--ta-label-color);
    }
  }

  &_link-icon {
    margin-left: 4px;

    --ta-icon-size: 20px;
    --ta-icon-color: var(--ta-content-color);

    &.arrow {
      &--down {
        transform: rotate(90deg);
      }

      &--left {
        transform: rotate(180deg);
      }

      &--up {
        transform: rotate(270deg);
      }
    }
  }

  &_body {
    width: 100%;
    line-height: 20px;
    font-size: 14px;
    color: $font2-color;
    margin: -8px 0 0 16px;
    padding-bottom: 12px;
    word-break: break-all;
    word-wrap: break-word;
    overflow: hidden;
    text-align: left;

    .ta-cell.has--icon & {
      margin-left: 48px;
    }
  }

  &:not(.disabled).clickable {
    cursor: pointer;

    &:active {
      background-color: $background-color;
    }
  }

  &.disabled {
    --ta-content-color: #{rgba(var(--ta-grey-8), 0.2)};

    .ta-cell {
      &_label,
      &_body {
        opacity: 0.2;
      }
    }

    cursor: not-allowed;
  }

  .ta {
    &-input {
      height: 32px;
      border-width: 0;
      padding: 4px 0;
      border-radius: 0;

      &_input {
        text-align: right;
      }

      &.disabled {
        background-color: transparent;
      }
    }

    &-textarea {
      height: 92px;
    }

    &-radio-group {
      justify-content: flex-end;

      &.vertical .ta-radio_box {
        justify-content: flex-end;
      }
    }

    &-checkbox-group {
      justify-content: flex-end;

      &.vertical .ta-checkbox_box {
        justify-content: flex-end;
      }
    }

    &-picker-view {
      margin: -8px 0;
    }

    &-image-uploader {
      padding: 0;
      width: 192px;
      margin-left: auto;

      &.column-2 {
        width: 128px;
      }

      &.column-1 {
        width: 64px;
      }
    }
  }
}
