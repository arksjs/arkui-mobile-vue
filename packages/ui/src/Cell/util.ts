import { getEnumsValue } from '../helpers'
import type { ArrowDirection, CellProps } from './types'

export const LINK_ICON_NAMES: ArrowDirection[] = ['right', 'up', 'down', 'left']

export const getCellClasses = (props: CellProps) => {
  return [
    'ta-cell',
    'ta-horizontal-hairline',
    {
      clickable: !!(props.clickable || props.isLink),
      'has--icon': props.icon,
      disabled: props.disabled
    }
  ]
}

export const getCellArrowClasses = (arrowDirection?: ArrowDirection) => {
  return ['ta-cell_link-icon', 'arrow--' + getEnumsValue(LINK_ICON_NAMES, arrowDirection)]
}
