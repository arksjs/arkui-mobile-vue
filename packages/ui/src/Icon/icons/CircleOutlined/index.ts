export { default as default } from './CircleOutlined.vue'
