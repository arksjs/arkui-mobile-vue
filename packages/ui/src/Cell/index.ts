import { withInstall } from '../helpers/with-install'
import Cell from './Cell.vue'

export { Cell }
export const install = withInstall(Cell)
export default Cell
