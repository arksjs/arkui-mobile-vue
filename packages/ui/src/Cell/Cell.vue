<template>
  <div :class="classes" @click="onClick">
    <div class="ak-cell_header">
      <div class="ak-cell_icon" v-if="$slots.icon">
        <slot name="icon"></slot>
      </div>
      <div class="ak-cell_icon" v-else-if="icon"><Icon :icon="icon" /></div>
      <div class="ak-cell_label" v-if="label">
        {{ label }}
        <span class="ak-cell_required" v-if="required">*</span>
      </div>
      <div class="ak-cell_content">
        <slot>{{ content }}</slot>
      </div>
      <Icon v-if="isLink" :class="arrowClasses" :icon="RightOutlined" />
    </div>
    <div class="ak-cell_body" v-if="description">
      {{ description }}
    </div>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue'
import type { PropType } from 'vue'
import { Icon } from '../Icon'
import {
  createEnumsValidator,
  emitClickValidator,
  iconValidator
} from '../helpers/validator'
import type { ArrowDirection, CellEmits } from './types'
import RightOutlined from '../Icon/icons/RightOutlined'
import type { IconData } from '../Icon/types'
import { getCellArrowClasses, getCellClasses, LINK_ICON_NAMES } from './util'
import type { OnClick, PropsToEmits } from '../helpers/types'

export default defineComponent({
  name: 'ak-cell',
  components: { Icon },
  props: {
    icon: {
      type: [String, Object] as PropType<IconData>,
      validator: iconValidator
    },
    label: {
      type: String,
      default: ''
    },
    description: {
      type: String,
      default: ''
    },
    content: {
      type: String,
      default: ''
    },
    clickable: {
      type: Boolean,
      default: false
    },
    required: {
      type: Boolean,
      default: false
    },
    isLink: {
      type: Boolean,
      default: false
    },
    arrowDirection: {
      type: String as PropType<ArrowDirection>,
      validator: createEnumsValidator(LINK_ICON_NAMES)
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    click: emitClickValidator
  } as PropsToEmits<CellEmits>,
  setup(props, { emit }) {
    const onClick: OnClick = e => {
      if (!props.disabled) {
        emit('click', e)
      }
    }

    const classes = computed(() => getCellClasses(props))
    const arrowClasses = computed(() =>
      getCellArrowClasses(props.arrowDirection)
    )

    return {
      classes,
      arrowClasses,
      RightOutlined,
      onClick
    }
  }
})
</script>
