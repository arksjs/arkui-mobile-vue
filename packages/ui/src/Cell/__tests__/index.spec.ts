import { describe, test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { Cell } from '@/Cell'
import CircleOutlined from '@/Icon/icons/CircleOutlined'

const label = 'label'
const content = 'content'
const description = 'description'

describe('Cell', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Cell, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('label & content & description', () => {
    const wrapper = shallowMount(Cell, {
      props: {
        label,
        content,
        description
      }
    })

    expect(wrapper.find('.ta-cell_label').text()).toBe(label)
    expect(wrapper.find('.ta-cell_content').text()).toBe(content)
    expect(wrapper.find('.ta-cell_body').text()).toBe(description)
  })

  test('clickable', async () => {
    const wrapper = mount(Cell, {
      props: {
        clickable: true
      }
    })

    expect(wrapper.classes()).toContain('clickable')
  })

  test('isLink & arrowDirection', async () => {
    const wrapper = mount(Cell, {
      props: {
        isLink: true
      }
    })

    // isLink
    expect(wrapper.find('.ta-cell_link-icon').exists()).toBeTruthy()
    expect(wrapper.find('.ta-cell_link-icon').classes()).toContain(
      'arrow--right'
    )

    // arrowDirection
    await wrapper.setProps({
      arrowDirection: 'up'
    })
    expect(wrapper.find('.ta-cell_link-icon').classes()).toContain('arrow--up')
    await wrapper.setProps({
      arrowDirection: 'down'
    })
    expect(wrapper.find('.ta-cell_link-icon').classes()).toContain(
      'arrow--down'
    )
    await wrapper.setProps({
      arrowDirection: 'left'
    })
    expect(wrapper.find('.ta-cell_link-icon').classes()).toContain(
      'arrow--left'
    )
  })

  test('slots', () => {
    const wrapper = shallowMount(Cell, {
      slots: {
        icon: CircleOutlined,
        default: content
      }
    })

    expect(wrapper.findComponent(CircleOutlined).exists()).toBeTruthy()
    expect(wrapper.find('.ta-cell_content').text()).toBe(content)
  })
})
