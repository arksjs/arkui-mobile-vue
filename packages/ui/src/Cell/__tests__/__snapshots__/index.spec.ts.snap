// Vitest Snapshot v1

exports[`Cell > snapshot 1`] = `
"<div class=\\"ta-cell ta-horizontal-hairline\\">
  <div class=\\"ta-cell_header\\">
    <!--v-if-->
    <!--v-if-->
    <div class=\\"ta-cell_content\\"></div>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>"
`;
